<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#11.Ridge,-Lasso,-Elastic-선형회귀분석-과제" data-toc-modified-id="11.Ridge,-Lasso,-Elastic-선형회귀분석-과제-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>11.Ridge, Lasso, Elastic 선형회귀분석 과제</a></span><ul class="toc-item"><li><span><a href="#전처리-:--id,-date-삭제" data-toc-modified-id="전처리-:--id,-date-삭제-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>전처리 :  id, date 삭제</a></span></li><li><span><a href="#범주형-변수-전처리" data-toc-modified-id="범주형-변수-전처리-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>범주형 변수 전처리</a></span></li><li><span><a href="#종속변수-y-=-price-,-나머지는-독립변수-X로-설정" data-toc-modified-id="종속변수-y-=-price-,-나머지는-독립변수-X로-설정-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>종속변수 y = price , 나머지는 독립변수 X로 설정</a></span></li><li><span><a href="#홀드아웃-기법을-사용하여-데이터-분할" data-toc-modified-id="홀드아웃-기법을-사용하여-데이터-분할-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>홀드아웃 기법을 사용하여 데이터 분할</a></span></li><li><span><a href="#Ridge-회귀분석을-하여-R2-값과-RMSE값을-구하고-설명하세요." data-toc-modified-id="Ridge-회귀분석을-하여-R2-값과-RMSE값을-구하고-설명하세요.-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Ridge 회귀분석을 하여 R2 값과 RMSE값을 구하고 설명하세요.</a></span><ul class="toc-item"><li><span><a href="#해석" data-toc-modified-id="해석-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>해석</a></span></li></ul></li><li><span><a href="#스케일링-한-데이터셋의-결과와-스케일링-하지-않은-데이터-셋의-결과를-비교해보세요." data-toc-modified-id="스케일링-한-데이터셋의-결과와-스케일링-하지-않은-데이터-셋의-결과를-비교해보세요.-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>스케일링 한 데이터셋의 결과와 스케일링 하지 않은 데이터 셋의 결과를 비교해보세요.</a></span><ul class="toc-item"><li><span><a href="#해석" data-toc-modified-id="해석-1.6.1"><span class="toc-item-num">1.6.1&nbsp;&nbsp;</span>해석</a></span></li></ul></li><li><span><a href="#회귀-계수-중-집값과-상관성이-높은-3가지-변수를-해석하세요" data-toc-modified-id="회귀-계수-중-집값과-상관성이-높은-3가지-변수를-해석하세요-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>회귀 계수 중 집값과 상관성이 높은 3가지 변수를 해석하세요</a></span><ul class="toc-item"><li><span><a href="#해석" data-toc-modified-id="해석-1.7.1"><span class="toc-item-num">1.7.1&nbsp;&nbsp;</span>해석</a></span></li></ul></li></ul></li></ul></div>

# 11.Ridge, Lasso, Elastic 선형회귀분석 과제 

변수 설명 

id - 판매된 각 주택의 고유 ID <br>
date  - 주택 판매 날짜<br>
price - 판매된 각 주택의 가격<br>
bedrooms  - 침실 수<br>
bathrooms  - 화장실이 있지만 샤워 시설은 없는 방을 0.5로 설명하는 욕실 수<br>
sqft_living - 아파트 내부 생활 공간의 평방 피트<br>
sqft_lot - 토지 공간의 평방 피트수<br>
floors  - 층 수<br>
waterfront - 아파트가 해안가를 내려다보고 있는지 여부에 대한 더미 변수<br>
view - 부동산에 대한 뷰가 얼마나 좋았는지에 대한 0에서 4 사이의 인덱스<br>
condition - 아파트 상태에 대한 1에서 5까지의 지수,<br>
grade - 1-13까지의 수로,1-3은 건축 및 설계 미달, 7은 시공 및 설계 평균, 11-13은 시공 및 설계 수준 높은 수준이다.<br>
sqft_above - 지면보다 높은 내부 주택 공간의 평방 피트수<br>
sqft_basement - 지면 아래에 있는 내부 주택 공간의 평방 피트수<br>
yr_built - 집이 처음 지어진 연도<br>
yr_renovated - 집의 마지막 리모델링 연도<br>
sqft_living15 - 가장 가까운 이웃 15명을 위한 내부 주택 생활 공간의 평방 피트<br>
sqft_lot15 - 가장 가까운 15개 이웃 토지 부지의 평방 피트수<br>

In [5]:
import pandas as pd 
df = pd.read_csv("./data/kc_house_data.csv")
df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors waterfront  view  condition  grade  \
0             1180      5650     1.0   standard     0          3      7   
1             2570      7242     2.0   standard     0          3      7   
2              770     10000     1.0   standard     0          3      6   
3             1960      5000     1.0   standard     0          5      7   
4             1680      8080     1.0   standard     0          3      8   
...            ...       ...     ...        ...   ...        ...    ...   
21608         1530      1131     3.0   standard     0          3      8   
21609         2310      5813     2.0   standard     0          3      8   
21610         1020      1350     2.0   standard     0          3      7   
21611         1600      2388     2.0   standard     0          3      8   
21612         1020      1076     2.0   standard     0          3      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  sqft_living15  \
0            1180              0      1955             0           1340   
1            2170            400      1951          1991           1690   
2             770              0      1933             0           2720   
3            1050            910      1965             0           1360   
4            1680              0      1987             0           1800   
...           ...            ...       ...           ...            ...   
21608        1530              0      2009             0           1530   
21609        2310              0      2014             0           1830   
21610        1020              0      2009             0           1020   
21611        1600              0      2004             0           1410   
21612        1020              0      2008             0           1020   

       sqft_lot15  
0            5650  
1            7639  
2            8062  
3            5000  
4            7503  
...           ...  
21608        1509  
21609        7200  
21610        2007  
21611        1287  
21612        1357  

[21613 rows x 18 columns]

## 전처리 :  id, date 삭제 

In [6]:
df.drop(columns=['id','date'], inplace=True)

## 범주형 변수 전처리 

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   price          21613 non-null  float64
 1   bedrooms       21613 non-null  int64  
 2   bathrooms      21613 non-null  float64
 3   sqft_living    21613 non-null  int64  
 4   sqft_lot       21613 non-null  int64  
 5   floors         21613 non-null  float64
 6   waterfront     21613 non-null  object 
 7   view           21613 non-null  int64  
 8   condition      21613 non-null  int64  
 9   grade          21613 non-null  int64  
 10  sqft_above     21613 non-null  int64  
 11  sqft_basement  21613 non-null  int64  
 12  yr_built       21613 non-null  int64  
 13  yr_renovated   21613 non-null  int64  
 14  sqft_living15  21613 non-null  int64  
 15  sqft_lot15     21613 non-null  int64  
dtypes: float64(3), int64(12), object(1)
memory usage: 2.6+ MB


In [8]:
df['waterfront'].value_counts()

standard      21450
river_view      163
Name: waterfront, dtype: int64

In [10]:
df_dum = pd.get_dummies(columns=['waterfront'], data = df)

## 종속변수 y = price , 나머지는 독립변수 X로 설정 

In [11]:
y = df_dum['price']
X = df_dum.drop(columns = ['price'])

## 홀드아웃 기법을 사용하여 데이터 분할 

In [16]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7,
                                                   test_size = 0.3, random_state=123)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(15129, 16) (6484, 16) (15129,) (6484,)


## Ridge 회귀분석을 하여 R2 값과 RMSE값을 구하고 설명하세요. 

In [17]:
from sklearn.linear_model import Ridge 
from sklearn.model_selection import GridSearchCV

ridge = Ridge() 
param_grid = {'alpha':(0.0001,0.0005,0.001,0.005, 0.01,0.05,0.1, 0.15, 0.2, 0.25, 0.3,0.5,1,2,3)}
model = GridSearchCV(ridge, param_grid)
model.fit(X_train,y_train)

/Users/hmc/opt/anaconda3/envs/ADP_Class/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=3.25626e-18): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, overwrite_a=True).T
/Users/hmc/opt/anaconda3/envs/ADP_Class/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=3.02031e-18): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, overwrite_a=True).T
/Users/hmc/opt/anaconda3/envs/ADP_Class/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=3.30285e-18): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, overwrite_a=True).T
/Users/hmc/opt/anaconda3/envs/ADP_Class/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=3.31546e-18): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, over

GridSearchCV(estimator=Ridge(),
             param_grid={'alpha': (0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05,
                                   0.1, 0.15, 0.2, 0.25, 0.3, 0.5, 1, 2, 3)})

In [18]:
model.best_estimator_

Ridge(alpha=2)

In [19]:
model = Ridge(alpha=2) 
model.fit(X_train,y_train)

Ridge(alpha=2)

In [21]:
## R2
import numpy as np
print("사용한 데이터 컬럼 수 :", np.sum(model.coef_!=0))
print("훈련 데이터 셋 성능 : ", model.score(X_train,y_train))
print("테스트 데이터 셋 성능 : ", model.score(X_test,y_test))

사용한 데이터 컬럼 수 : 16
훈련 데이터 셋 성능 :  0.6494005566522837
테스트 데이터 셋 성능 :  0.6629505285083124


In [25]:
from sklearn.metrics import mean_squared_error
y_pred = model.predict(X_test)

mse = mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)
print(rmse)

213942.0201695023


In [26]:
y_pred

array([618321.5705133 , 400302.93179583, 909678.4501467 , ...,
       858556.1105221 , 581383.25813401, 469600.98007551])

### 해석
- alpha 값이 (0.0001,0.0005,0.001,0.005, 0.01,0.05,0.1, 0.15, 0.2, 0.25, 0.3,0.5,1,2,3) 사이에서 2가 최적인 것으로 확인
- 이때, 사용된 종속 변수는 16개임
- R2_score_train = 0.65
- R2_score_test = 0.66
- RMSE 값은 213942

## 스케일링 한 데이터셋의 결과와 스케일링 하지 않은 데이터 셋의 결과를 비교해보세요.
> Tip : X_train 데이터를 fit_transform 하여 전처리, X_test는 transform 사용하여 전처리 <br>
>  X_train_scaled = pd.DataFrame(scaler.fit_transform(X_train),columns=X_train.columns,index=X_train.index)

In [49]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = pd.DataFrame(scaler.fit_transform(X_train),columns=X_train.columns,index=X_train.index)
X_test_scaled = pd.DataFrame(scaler.transform(X_test),columns=X_test.columns,index=X_test.index)

In [50]:
X_test_scaled.shape

(6484, 16)

In [51]:
X_train_scaled.shape

(15129, 16)

In [52]:
ridge = Ridge() 
param_grid = {'alpha':(0.0001,0.0005,0.001,0.005, 0.01,0.05,0.1, 0.15, 0.2, 0.25, 0.3,0.5,1,2,3)}
model = GridSearchCV(ridge, param_grid)
model.fit(X_train_scaled,y_train)

GridSearchCV(estimator=Ridge(),
             param_grid={'alpha': (0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05,
                                   0.1, 0.15, 0.2, 0.25, 0.3, 0.5, 1, 2, 3)})

In [53]:
model.best_estimator_

Ridge(alpha=3)

In [54]:
model = Ridge(alpha=3)
model.fit(X_train_scaled,y_train)

Ridge(alpha=3)

In [55]:
## R2
import numpy as np
print("사용한 데이터 컬럼 수 :", np.sum(model.coef_!=0))
print("훈련 데이터 셋 성능 : ", model.score(X_train_scaled,y_train))
print("테스트 데이터 셋 성능 : ", model.score(X_test_scaled,y_test))
y_pred = model.predict(X_test_scaled)

mse = mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)
print("RMSE :",rmse)

사용한 데이터 컬럼 수 : 16
훈련 데이터 셋 성능 :  0.6494027213510856
테스트 데이터 셋 성능 :  0.6628356990191566
RMSE : 213978.46105266176


### 해석
- alpha 값이 (0.0001,0.0005,0.001,0.005, 0.01,0.05,0.1, 0.15, 0.2, 0.25, 0.3,0.5,1,2,3) 중 Ridge 회귀했을 때 alpha=3이 최적임을 확인
- 회귀에 사용된 종속변수는 16개임
- R2_score_train = 0.65
- R2_score_test = 0.66
- RMSE = 213978
- StandardScaled 된 결과와 비교했을 때, Ridge 회귀의 매개변수는 변했지만(alpha), R2_score와 RMSE는 거의 변함이 없다.

## 회귀 계수 중 집값과 상관성이 높은 3가지 변수를 해석하세요 

In [56]:
model.coef_

array([ -34146.27927065,   34044.02924205,   74077.50173883,
           646.47835578,   13250.7523758 ,   31542.0209924 ,
         13025.70233902,  145101.6389495 ,   62835.31126402,
         36028.01209128, -106714.07707908,    3780.09355268,
         15741.59833662,  -15494.79208757,   26341.43981365,
        -26341.43981366])

In [71]:
coefficients = pd.concat([pd.DataFrame(model.coef_), pd.DataFrame(model.feature_names_in_)], axis = 1)
coefficients.columns = ['coef','feature']

In [76]:
coefficients

coef                feature
0   -34146.279271               bedrooms
1    34044.029242              bathrooms
2    74077.501739            sqft_living
3      646.478356               sqft_lot
4    13250.752376                 floors
5    31542.020992                   view
6    13025.702339              condition
7   145101.638950                  grade
8    62835.311264             sqft_above
9    36028.012091          sqft_basement
10 -106714.077079               yr_built
11    3780.093553           yr_renovated
12   15741.598337          sqft_living15
13  -15494.792088             sqft_lot15
14   26341.439814  waterfront_river_view
15  -26341.439814    waterfront_standard

In [80]:
coefficients['abs_coef'] = abs(coefficients['coef'])

In [81]:
coefficients

coef                feature       abs_coef
0   -34146.279271               bedrooms   34146.279271
1    34044.029242              bathrooms   34044.029242
2    74077.501739            sqft_living   74077.501739
3      646.478356               sqft_lot     646.478356
4    13250.752376                 floors   13250.752376
5    31542.020992                   view   31542.020992
6    13025.702339              condition   13025.702339
7   145101.638950                  grade  145101.638950
8    62835.311264             sqft_above   62835.311264
9    36028.012091          sqft_basement   36028.012091
10 -106714.077079               yr_built  106714.077079
11    3780.093553           yr_renovated    3780.093553
12   15741.598337          sqft_living15   15741.598337
13  -15494.792088             sqft_lot15   15494.792088
14   26341.439814  waterfront_river_view   26341.439814
15  -26341.439814    waterfront_standard   26341.439814

In [82]:
coefficients.sort_values(by=['abs_coef'], ascending=False)[0:3]

coef      feature       abs_coef
7   145101.638950        grade  145101.638950
10 -106714.077079     yr_built  106714.077079
2    74077.501739  sqft_living   74077.501739

### 해석
- 종속변수 중 grade, yr_built, sqft_living 변수가 독립변수 price에 대해 가장 높은 계수의 절대값을 갖고 있어 영향도가 높다고 할 수 있음
- grade 가 커질 수록 price도 커짐
- yr_built 가 커질 수록 price는 작아짐
- sqft_living 가 커질 수록 price도 커짐